### PR TITLE
Add a stable version of Anderson mixing

### PR DESCRIPTION
**Pull request by haampie**
and a mini app for testing purposes. Uses repeated Gram-Schmidt to
append columns and Given's rotations to drop columns.

This is a stable version of what is called broyden1 in the code,
but that's a bit of a misnomer, since our broyden1 is Anderson aka
a member of the broyden2 family.

The mini app solves this fixed-point problem g(x) = x, with g(x) = Ax - b where A is a diagonal matrix with `A(i, i) = 2 + 1 / i`. So effectively solving a linear equation (A - I)x = b. The method is equivalent to our current broyden1 method, but should be more stable, since it does explicitly creates (ΔFᵀΔF)⁻¹ but builds a QR-factorization of ΔF on the go by appending and dropping columns. This seems to improve the number of iterations sometimes a bit, when solving to a low tolerance:

```
$ ./test_anderson --max_history=4 --beta=0.15 --dim=100 --max_iter=100 --mixer=broyden1 --tol=1e-13
max history = 4. beta = 0.15. dim = 100. mixer = broyden1
    iter                       ||res||                       ||err||
       0              10.5834653557077                            10
       1              12.3118067791144              11.5791378082023
       2              2.37025102260193              1.48393276943419
       3             0.321327286201523             0.246951057698304
       4            0.0401066072844496            0.0328090507049631
       5             0.047643496783862            0.0390244758264859
       6           0.00470684758111809           0.00397150164321658
       7           0.00107255095204715          0.000692588824346489
       8           0.00134692757909549          0.000860346927362686
       9          0.000246145351134748          0.000215808030720198
      10          3.29281396236896e-05          2.82342901561596e-05
      11          3.88686477211914e-05          3.32642753713581e-05
      12          4.93928028553194e-06          3.71216412039638e-06
      13          1.00338614240318e-06           6.2171188626742e-07
      14          1.27071314055913e-06          7.70040449278316e-07
      15          1.99046636640774e-07          1.75871265369531e-07
      16           3.4786545623417e-08          2.91401766708109e-08
      17          4.12080157326562e-08           3.4328232983216e-08
      18           5.2072569502646e-09          4.36045555915723e-09
      19          1.25477380275483e-09          8.30355017802606e-10
      20          1.57116623049378e-09          1.02501357231777e-09
      21          3.16309750278533e-10            2.901368902634e-10
      22          5.26858719464967e-11          4.41799325233912e-11
      23          6.24341371178346e-11          5.23933563078012e-11
      24          7.90966323856929e-12          6.35911034999388e-12
      25          2.32045833720214e-12          1.55230892587231e-12
      26          2.88780734957512e-12          1.93991026583153e-12
      27          5.02483886964201e-13          4.25255172992851e-13
      28          1.89133392316426e-13          1.50743758467021e-13
      29          2.25731030256117e-13          1.79065172985851e-13
      30          3.06012993883521e-14          2.57085560292118e-14
$ ./test_anderson --max_history=4 --beta=0.15 --dim=100 --max_iter=100 --mixer=stable_anderson --tol=1e-13
max history = 4. beta = 0.15. dim = 100. mixer = stable_anderson
    iter                       ||res||                       ||err||
       0              10.5834653557077                            10
       1              12.3118067791144              11.5791378082023
       2              2.37025102260192              1.48393276943419
       3             0.321327286201433              0.24695105769854
       4             0.040106607266002            0.0328090507738832
       5            0.0476434967838501            0.0390244758264763
       6            0.0047068475811195           0.00397150164321969
       7           0.00107255044179766          0.000692588526603841
       8           0.00134692693294447          0.000860346544196874
       9          0.000246145277281368          0.000215807958219579
      10          3.29281135255153e-05          2.82342673760101e-05
      11           3.8868617036493e-05          3.32642485787986e-05
      12          4.93927746400295e-06          3.71216216310097e-06
      13          1.00338494060677e-06          6.21711235830479e-07
      14          1.27071158758686e-06          7.70039626995564e-07
      15          1.99046483218068e-07          1.75871124093968e-07
      16          3.47861643479396e-08          2.91398671744918e-08
      17          4.12075624707466e-08          3.43278651516252e-08
      18          5.20718558974553e-09          4.36039249563765e-09
      19          1.25483322592963e-09          8.30390328646485e-10
      20          1.57124261815548e-09           1.0250580554414e-09
      21          3.16319761371464e-10          2.90148200490116e-10
      22          5.28052739847229e-11           4.4281333405316e-11
      23           6.2575755410084e-11           5.2512350572028e-11
      24          7.92171512138448e-12          6.36848042560268e-12
      25          2.56183259268808e-12          1.71197687812445e-12
      26          3.18719996003772e-12          2.14068238293792e-12
      27          5.23477632108497e-13          4.40191038832539e-13
      28          9.40379873959412e-14          7.69807381984938e-14
```

Have to test this on non-mini-app problems still!

### CODE REVIEW

**Comment by haampie** [comment]
@electronic-structure run

```yaml
spec: sirius@develop +scalapack

current:
  cmd: sirius.scf --mixer.type=stable_anderson
```

**Comment by electronic-structure** [comment]
<details>
<summary>Benchmark <code>benchmarks/5172a58</code> submitted:</summary>

```json
{
    "id": "6ab383ce-5393-42c2-a618-2383cdb0465c",
    "reference": {
        "spec": "sirius@develop +scalapack",
        "cmd": [
            "sirius.scf"
        ],
        "repo": "electronic-structure/SIRIUS",
        "sha": "0c261670dec9b96a58dd74a9134f19b00042d6d1",
        "build": true
    },
    "current": {
        "spec": "sirius@develop +scalapack",
        "cmd": [
            "sirius.scf",
            "--mixer.type=stable_anderson"
        ],
        "repo": "haampie/SIRIUS",
        "sha": "4e60b1847c4444fb88c162c63a530351544e3d41",
        "build": true
    },
    "report_to": {
        "repository": "electronic-structure/SIRIUS",
        "type": "pr",
        "issue": 584
    }
}

```
</details>

To trigger benchmarks again using cached binaries, include the following in your comment:
```yaml
id: 6ab383ce-5393-42c2-a618-2383cdb0465c
```

**Comment by electronic-structure** [comment]
Report for `benchmarks/5172a587` at https://github.com/electronic-structure/SIRIUS-bench-reports/tree/master/2020-11/11/4e60b18_vs_0c26167-5172a587/

```diff
  Si63Ge ───────────────────  72.58 s  →  72.43 s   (-0.21%)
  Au-surf ────────────────── 120.63 s 
  SiO2-ANA-kpoint-parallel ─ 132.74 s  → 133.18 s   (+0.33%)
  SiO2-ANA-band-parallel ─── 364.48 s  → 349.16 s   (-4.20%)
```

**Comment by haampie** [comment]
This should close #539 for the old broyden1 mixer as well now. Whenever there's an exact breakdown because the least-squares problem has no solution (when ΔFᵀΔF is not invertible), we simply restart. In the step of the breakdown, it just performs a linear mixing step.

**Comment by haampie** [comment]
bors try

File: src/mixer/anderson_mixer.hpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+/** \file anderson_mixer.hpp
+ *
+ *   \brief Contains definition and implementation sirius::Anderson.
+ */
+
+#ifndef __Anderson_MIXER_HPP__
```
Comment (toxa81) [inline src/mixer/anderson_mixer.hpp:25]:
```
#ifdef __ANDERSON_MIXER_HPP__
```

File: src/mixer/anderson_mixer.hpp [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+#include "SDDK/memory.hpp"
+#include "mixer/mixer.hpp"
+#include "linalg/linalg.hpp"
+
+namespace sirius {
+namespace mixer {
+
+/// Anderson mixer.
+/**
```
Comment (toxa81) [inline src/mixer/anderson_mixer.hpp:46]: This is a Doxygen comment, which allows for  LaTeX formulas. I would prefer to not have UTF characters here

File: src/mixer/anderson_stable_mixer.hpp [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+ * iterations. Then Gₙ is taken -βI. This implementation uses Gram-Schmidt
+ * to orthogonalize ΔFₙ to solve the least-squares problem. If stability is
+ * not an issue, use Anderson instead.
+ * 
+ * Reference paper: Fang, Haw‐ren, and Yousef Saad. "Two classes of multisecant
+ * methods for nonlinear acceleration." Numerical Linear Algebra with Applications
+ * 16.3 (2009): 197-221.
+ */
+template <typename... FUNCS>
+class AndersonStable : public Mixer<FUNCS...>
```
Comment (toxa81) [inline src/mixer/anderson_stable_mixer.hpp:63]: SIRIUS has a funny naming convention
```
Anderson_stable
```

**Comment by toxa81** [comment]
```
-DCREATE_PYTHON_MODULE=1
```
is not tested. Python module is broken at the moment

**Pull request merged**